tutorial/practicals/LFRic/building_code/3_time_evolution/solutions/prop_perturbation_kernel_mod.f90

```diff
@@ -0,0 +1,140 @@
+!                      exp( -((x - x_centre - u_vel*t_tot)/half_width_x)**2 &
+!                           -((y - y_centre - v_vel*t_tot)/half_width_y)**2 )
+! -----------------------------------------------------------------------------
+module prop_perturbation_kernel_mod
+
+  use argument_mod,      only: arg_type, func_type,   &
+                               GH_FIELD, GH_REAL,     &
+                               GH_READWRITE, GH_READ, &
+                               CELLS
+  use fs_continuity_mod, only: W3
+  use constants_mod,     only: r_def, i_def
+  use kernel_mod,        only: kernel_type
+  use perturbation_bell_config_mod, &
+                         only: half_width_x, half_width_y, &
+                               perturbation_scale,         &
+                               perturbation_height,        &
+                               x_centre, y_centre, u_vel, v_vel
+
+  implicit none
+
+  private
+
+  !-----------------------------------------------------------------------------
+  ! The type declaration for the kernel. Contains the metadata needed by
+  ! the PSy layer.
+  !-----------------------------------------------------------------------------
+  type, public, extends(kernel_type) :: prop_perturbation_kernel_type
+    private
+    type(arg_type), dimension(3) :: meta_args = (/ &
+         arg_type(GH_FIELD,   GH_READWRITE, W3),   &
+         arg_type(GH_FIELD*3, GH_READ,      W3),   &
+         arg_type(GH_REAL,    GH_READ)             &
+         /)
+    integer :: iterates_over = CELLS
+  contains
+    procedure, nopass :: prop_perturbation_code
+  end type prop_perturbation_kernel_type
+
+  public prop_perturbation_code
+
+  contains
+
+  !> @brief Initialise a perturbation field using coordinate and namelist data
+  !> @param[in] nlayers Number of layers
+  !> @param[in,out] perturbation Perturbation field to propagate
+  !> @param[in] chi_1 Coordinates in the x direction
+  !> @param[in] chi_2 Coordinates in the y direction
+  !> @param[in] chi_3 Coordinates in the z direction
+  !> @param[in] t_tot Time (s) to propagate the perturbation field for
+  !> @param[in] ndf_w3 Number of degrees of freedom per cell for the
+  !!                   perturbation field
+  !> @param[in] undf_w3 Number of unique degrees of freedom for the
+  !!                   perturbation field
+  !> @param[in] map_w3 Dofmap for the cell at the base of the column for the
+  !!                   perturbation field
+  subroutine prop_perturbation_code(nlayers, perturbation,      &
+                                    chi_1, chi_2, chi_3, t_tot, &
+                                    ndf_w3, undf_w3, map_w3)
+
+    implicit none
+
+    ! Arguments
+    integer(kind=i_def), intent(in) :: nlayers
+    integer(kind=i_def), intent(in) :: ndf_w3
+    integer(kind=i_def), intent(in) :: undf_w3
+    integer(kind=i_def), intent(in), dimension(ndf_w3) :: map_w3
+    real(kind=r_def), intent(inout), dimension(undf_w3) :: perturbation
```

Comment:
Perturbation is declared as GH_READWRITE, and intent(inout) - but as far as I can tell it is only written to, not read.

Reply:
This should actually be `write` whether `init_perturbation` is `readwrite` as there is a built-in call beforehand. Changed,